BHoM_UI/Components/UI/UnitTest.cs

```diff
@@ -0,0 +1,164 @@
+        /**** Override Methods            ****/
+        /*************************************/
+
+        protected override void SetOutputs(MethodBase method)
+        {
+            if (method == null)
+                OutputParams = new List<ParamInfo>();
+
+            OutputParams = new List<ParamInfo>();
+            OutputParams.Add(new ParamInfo
+            {
+                Name = "UnitTest",
+                DataType = typeof(UnitTest),
+                Description = "UnitTest for the method " + method.Name +".",
+                Kind = ParamKind.Output,
+                IsRequired = true
+            });
+        }
+
+        /*************************************/
+
+        protected override void SetComponentDetails(MethodBase method)
+        {
+            if (method == null)
+                return;
+
+            // Set component name
+            if (method is MethodInfo)
+                Name = "UT:" + method.Name;
+            else if (method is ConstructorInfo)
+                Name = "UT:" + method.DeclaringType.Name;
+            else
+                Name = "UT:UnknownMethod";
+
+            // Set description
+            Description = "Creates a UnitTest for the method " + method.Name + ".";
+        }
+
+        /*************************************/
+
+        public override object Run(List<object> inputs)
+        {
+            if (m_CompiledFunc != null)
+            {
+                object returnValue = null;
+                MethodInfo method = m_OriginalItem as MethodInfo;
+                try
+                {
+                    returnValue = m_CompiledFunc(inputs.ToArray());
+                }
+                catch (InvalidCastException e)
+                {
+                    MethodInfo originalMethod = m_OriginalItem as MethodInfo;
+                    if (originalMethod != null && originalMethod.IsGenericMethod)
+                    {
+                        // Try to update the generic method to fit the input types
+                        method = Engine.Reflection.Compute.MakeGenericFromInputs(originalMethod, inputs.Select(x => x.GetType()).ToList());
+                        m_CompiledFunc = method.ToFunc();
+                        returnValue = m_CompiledFunc(inputs.ToArray());
+                    }
+                    else
+                        throw e;
+                }
+
+                return new UnitTest() { Method = method, Data = new List<TestData>() { new TestData(inputs, SeparateOutputs(returnValue)) } };
+            }
+            else if (InputParams.Count <= 0)
+            {
+                BH.Engine.Reflection.Compute.RecordWarning("This is a magic component. Right click on it and <Select a method>");
+                return null;
+            }
+            else
+            {
+                BH.Engine.Reflection.Compute.RecordError("The component is not linked to a method.");
+                return null;
+            }
+        }
```

Comment:
As requested previously, call the base method instead of just copy pasting the whole thing. This creates duplicates in the code that make it difficult to maintain. 

Here's an example of what you could do instead:

```c#
public override object Run(List<object> inputs)
{
    object returnValue = base.Run(inputs);

    if (m_CompiledFunc != null)
        return new UnitTest() { Method = m_OriginalItem as MethodInfo, Data = new List<TestData>() { new TestData(inputs, SeparateOutputs(returnValue)) } };
    else
        return null;
}
```

Reply:
Agreed simpler. That case does not capture the case with generic methods though, but maybe that is something to be resolved at the run side of the unit test rather than here.

Will update.